### ISSUE TITLE
`modernize-use-trailing-return-type` reports underlying type in fix-it for lamdba

### ISSUE THREAD

**Issue by firewave** [llvm/llvm-project#144510]
```cpp
#include <algorithm>
#include <cstddef>

void f()
{
    auto val = [](size_t m) {
        size_t a = 0;
        return std::max(m, a);
    };
}
```

```
<source>:6:16: warning: use a trailing return type for this lambda [modernize-use-trailing-return-type]
    6 |     auto val = [](size_t m) {
      |                ^
      |                             -> unsigned long
```

https://godbolt.org/z/9K4M7Mz8E

It should be reporting `size_t` instead. It seems to be related to the template call involved.

**Comment by vbvictor** [comment]
Seems like a wrong printing policy here:
https://github.com/llvm/llvm-project/blob/549bc55cc39bb9fb22df464bcf3b7d4d4a5ff507/clang-tools-extra/clang-tidy/modernize/UseTrailingReturnTypeCheck.cpp#L636
Setting `PrintAsCanonical` to `false` doesn't help immediately, probably when we use `Lambda->getCallOperator()` we get `CXXConversionDecl` that thinks it has `unsigned long` as a return type.

**Comment by firewave** [comment]
The same is reported with libc++: https://godbolt.org/z/Tda4vq17d

**Comment by vbvictor** [comment]
I couldn't find a way to fix this issue, It seems that `std::max` is instantiated with `unsigned long` even if `size_t` explicitly written like `std::max<size_t>(m, a)` https://godbolt.org/z/nna5aEcT7.
Note that the check already preserve type correctly for "easy case" where we directly return variable of type `size_t`: https://godbolt.org/z/4EsnGh1z8.

With `return std::max;` we have `CXXMethodDecl <col:27, line:14:5> line:11:16 constexpr operator() 'unsigned long (size_t) const' inline` https://godbolt.org/z/9z83TTc87
With `return a;` we have `CXXMethodDecl <col:27, line:9:5> line:6:16 constexpr operator() 'size_t (size_t) const' inline` https://godbolt.org/z/jTeK3eT1P.

It seems like clang in general don't preserve sugared types here, so the check can't do anything about it.
@mizvekov, do you think this is an issue in clang or there is a way to get the sugared type?

**Comment by mizvekov** [comment]
This probably needs template specialization resugaring in order to get the sugared type, which is something I have been working on to bring to clang.

But I will double check next week if my patches already take care of that.

**Comment by vbvictor** [comment]
Thank you, much appreciated!

**Comment by mizvekov** [comment]
Yeah, you can see the difference template specialization resugaring makes here: https://godbolt.org/z/eqKqWx64o

Note that godbolt seems to run clang-tidy built from a different branch: https://godbolt.org/z/P9cje1hvb

Testing this locally, the `modernize-use-trailing-return-type` check does not even fire. This could be because the branch is a bit old by this point. It will be a while before I rebase it again.
